NDTensors/src/blocksparse/blocksparse.jl

```diff
@@ -51,16 +41,41 @@ end
 function BlockSparse(
   ::Type{ElT}, ::UndefInitializer, blockoffsets::BlockOffsets, dim::Integer; vargs...
 ) where {ElT<:Number}
-  return BlockSparse(Vector{ElT}(undef, dim), blockoffsets; vargs...)
+  return BlockSparse(default_datatype(ElT)(undef, dim), blockoffsets; vargs...)
+end
+
+function BlockSparse(::UndefInitializer, blockoffsets::BlockOffsets, dim::Integer; vargs...)
+  return BlockSparse(NDTensors.default_eltype(), undef, blockoffsets, dim; vargs...)
 end
 
 function BlockSparse(blockoffsets::BlockOffsets, dim::Integer; vargs...)
-  return BlockSparse(Float64, blockoffsets, dim; vargs...)
+  elt = default_eltype()
+  datat = default_datatype(elt)
+  N = ndims(datat)
+  return BlockSparse(UnallocatedZeros{elt,N,datat}, blockoffsets, dim; vargs...)
 end
 
-function BlockSparse(::UndefInitializer, blockoffsets::BlockOffsets, dim::Integer; vargs...)
-  return BlockSparse(Float64, undef, blockoffsets, dim; vargs...)
+# TODO: Write as `(::Type{<:BlockSparse})()`.
+BlockSparse{ElT,DataT,N}() where {ElT,DataT,N} = BlockSparse(DataT(), BlockOffsets{N}())
+
+## End BlockSparse constructors
+
+## Blocksparse fieldtypes
+# TODO: Implement as `fieldtype(storagetype, :data)`.
+datatype(::Type{<:BlockSparse{<:Any,DataT}}) where {DataT} = DataT
+# TODO: Implement as `ndims(blockoffsetstype(storagetype))`.
+ndims(storagetype::Type{<:BlockSparse{<:Any,<:Any,N}}) where {N} = N
+# TODO: Implement as `fieldtype(storagetype, :blockoffsets)`.
+blockoffsetstype(storagetype::Type{<:BlockSparse}) = BlockOffsets{ndims(storagetype)}
+
+function set_datatype(storagetype::Type{<:BlockSparse}, datatype::Type{<:AbstractVector})
+  return BlockSparse{eltype(datatype),datatype,ndims(storagetype)}
+end
+
+function set_ndims(storagetype::Type{<:BlockSparse}, ndims)
+  return BlockSparse{eltype(storagetype),datatype(storagetype),ndims}
```

Comment:
Reminder to use `get_parameters` and `set_parameters` from `SetParameters` here. If we leave that for a future PR, I think we should leave comments about that as a reminder for later.

Reply:
This is good to remember. I think I could incorporate them in this PR. I will work on this tomorrow